### ISSUE TITLE
[Blocks] Component labels don't show in Blocks

### ISSUE THREAD

**Issue by abidlabs** [gradio-app/gradio#704]
Is this intentional? It makes hard to designate what each input and output is for. For example, I'd like to say that one slider is for speed and the other for angle:

![image](https://user-images.githubusercontent.com/1778297/155164725-1aeff7a2-495c-4d1f-a4bb-9dfb3d79a703.png)

Code to reproduce:

```python
import gradio as gr
import numpy as np
import random
import matplotlib.pyplot as plt

def generate(s, a):
    x = np.arange(-100,100,1)
    y = -x**2
    plt.plot(x, y)
    return plt

block = gr.Blocks()

with block:
    gr.static.Markdown("Let's do some kinematics. Plot the Trajectory of a Parabola based on its speed and angle!")
    output = gr.outputs.Image(type="plot")

    with gr.Row():
        speed = gr.inputs.Slider(label="Speed")
        angle = gr.inputs.Slider(label="Angle")
    
    btn = gr.Button(label="Run")
    btn.click(generate, [speed, angle], output)
    
block.launch()
```

**Comment by aliabid94** [comment]
yes it was intentional, seemed like labels were more relevant for Interfaces. We can have a constructor parameter `show_labels` which can toggle visibility of labels though.

**Comment by abidlabs** [comment]
My sense is that labels should be shown by default. At least in the interfaces that I’ve been playing around with, it makes sense to have the labels. Perhaps if the label is set to None (or not provided), then it should just be hidden?

**Comment by pngwn** [comment]
I think that this depends on the component. Previosuly we had labels for everything even if it didn't really make sense in the context of the component but when we removed them, we also removed it from components where it _does_ make sense.

Inputs are one example of a component that needs labels but they should be part of the component itself, rather than external labels.

Short version: Needs assessing on a component by component basis, for inputs we should add them. This is necessary for accessibility reasons anyway.

**Comment by abidlabs** [comment]
Input and output components both should have labels, right? At least that's what do we for `Interface`, and it is usually helpful. 

For what components does it not make sense to have associated labels?

**Comment by pngwn** [comment]
I think there is a risk that the labels could interfer with the layout the user wants. Take markdown, html, json, they won't always need label because the purpose may be obvious from the context. Someone could be writing an article using blocks for example. Markdown will never need a label imo, you can write markdown. Even components like image won't always need a label.

I think we can add optional labels to some of these components but we need to consider how that will interact with the layout a user is using, or come up with a design to allow labels to exist _within_ the component. I think Victor's design will facilitate these kinds of labels/ titles for the more 'widgety' components. 

So my thinking is:
- some components (like forms) should _always_ have labels or they are unusable and inaccessible.
- some components should have optional labels (most: images, charts)
- some components should never have labels (html, markdown)

I think we should also introduce the concept of captions for certain components, specifically images, charts, etc. You may have components that don't need a titure but need a caption to elaborate, think figures with a technical article.

**Comment by pngwn** [comment]
I do think labels make sense for interfaces though, yes. It is a constrained use-case and the labels make sense in that context. Victors new designs have some kind of label in place.

**Comment by abidlabs** [comment]
Yes, to me these seem like sensible defaults in each case. But wouldn't it be (1) easier to implement (2) more flexible for the user if we allowed (but did not require) labels in all of these cases? 

Because I can see exceptions:

For example:
> some components should never have labels (html, markdown)
* Except if `HTML` or `Markdown` is used as an output component in an `Interface`, in which case someone may want to pass in a label to designate what the output is. They _could_ do it in HTML/Markdown itself, but it would be much easier to just pass in a label and let Gradio handle it.
> some components (like forms) should always have labels or they are unusable and inaccessible
* Except if someone is making a chatbot, under which is an input text field. Someone might prefer to use a "placeholder" instead of a "label" to designate what the chatbot is for to lead to a cleaner UI.

So the flexible approach that I'm suggesting is that by default, `label=None`, in which case no label shows up. If the user wants a label, they simply provide one. For the `Interface` class, labels are automatically generated (e.g. based on parameter names, which we already do), or the user can override them.

Regarding the other point:

> come up with a design to allow labels to exist within the component

I'm conflicted about this. I can see how this might lead to a cleaner/simpler design. But we have interfaces that take in multiple kinds of inputs. For example, this one that has both images and sliders. And I think it would lead to inconsistent design if some components (like images) had labels within the component and some (like sliders) that I assume would have to have the label outside of the widget

<img width="472" alt="image" src="https://user-images.githubusercontent.com/1778297/157871921-6fa80172-336c-4229-8f92-07ef5c67214a.png">

**Comment by FarukOzderim** [comment]
> So the flexible approach that I'm suggesting is that by default, for all components, label=None, in which case no label shows up.
I think this solves the problem very easily.

**Comment by abidlabs** [comment]
This looks like its fixed in the latest redesign @pngwn. Keeping open just because I think you wanted to add labels to a few more components?

**Comment by abidlabs** [comment]
Testing this more extensively, I noticed a couple of issues:

1. When `label` is `None`, the label still shows up as null. 
2. For `Textbox`, the `label` shows up for output textboxes, but not input textboxes?

See here:

<img width="708" alt="image" src="https://user-images.githubusercontent.com/1778297/161818067-4b219393-1076-4320-9560-64db59838b02.png">

**Comment by pngwn** [comment]
Forms _always_ need a label. They are completely inaccessible otherwise. I think it should be a required parameter.

Output textboxes haven't been updated along with the input, I'll fix that.

**Comment by abidlabs** [comment]
When you say that form elements are inaccessible, do you mean from a technical point of view, or from a human accessibility perspective? If it's the latter, can't accessibility be provided through other ways, for example, a user may want to provide explanatory text before the textbox using Markdown instead of using our Textbox's label, which has a very specific look and feel that the user might not want?

**Comment by pngwn** [comment]
@abidlabs I know we just spoke about this but think more: for the output components, I think labels could be optional because they are mostly visual anyway. But for inputs the accessibility stuff I mentioned applies.

**Comment by pngwn** [comment]
Closed by #929.

Will open a more specific issue for the optional labels thing, so we can figure out exactl;y what we want to do.
